ca/certificate-authority.go

```diff
@@ -362,18 +406,56 @@ func (ca *CertificateAuthorityImpl) IssueCertificate(csr x509.CertificateRequest
 		ca.log.Notice(message)
 	}
 
-	// Convert the CSR to PEM
-	csrPEM := string(pem.EncodeToMemory(&pem.Block{
-		Type:  "CERTIFICATE REQUEST",
-		Bytes: csr.Raw,
-	}))
+	// At this point, we are willing to issue, so create a pending
+	// certificate request.
+	pendingCert := core.Certificate{RegistrationID: regID}
+	pendingCert, err = ca.SA.NewPendingCertificate(pendingCert)
+	if err != nil {
+		ca.log.AuditErr(err)
+		return emptyCert, err
+	}
+
+	// Dispatch signing request, or fail if the queue is full
+	request := signingRequest{
+		logEventID:     logEventID,
+		cert:           pendingCert,
+		csr:            csr,
+		commonName:     commonName,
+		hostNames:      hostNames,
+		earliestExpiry: earliestExpiry,
+	}
+	select {
+	case ca.signingQueue <- request:
+		// No problem; request is now in the queue
+		logEventResult = "successful"
+		return pendingCert, nil
+	default:
+		err = core.ServiceUnavailableError("CA signing queue is full.")
+		ca.log.AuditErr(err)
+		return emptyCert, err
+	}
+}
+
+func (ca *CertificateAuthorityImpl) sign(sr signingRequest) {
+	logEventResult := "error"
+	logEvent := blog.CertificateRequestEvent{
+		ID:            sr.logEventID + ".sign",
+		RequestMethod: "online",
+		RequestTime:   time.Now(),
+	}
+
+	// No matter what, log the request
+	defer func() {
+		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
+		ca.log.AuditObject(fmt.Sprintf("Certificate request - %s", logEventResult), logEvent)
```

Comment:
Do we need different audit tags besides "Certificate request"?  Maybe not, given that we can distinguish based on the ID suffixes?